### PR TITLE
Multiple Scattering Verification with TestEM15

### PR DESCRIPTION
**Pull request by whokion**
This MR includes an extension of multiple scattering verification with respect to TestEM15 (Geant4 extended electromagnetic example) and fixed a couple of bugs in skipping sampling condition and ordering angular sampling.

### CODE REVIEW

File: src/celeritas/em/distribution/UrbanMscStepLimit.hh [hunk 1]
```diff
@@ -143,9 +143,6 @@ UrbanMscStepLimit::UrbanMscStepLimit(const UrbanMscRef&       shared,
     range_ = helper_.range();
     // Mean free path for MSC at current energy
     lambda_ = helper_.msc_mfp(inc_energy_);
-
-    // The slowing down range should already have been applied as a step limit
-    CELER_ENSURE(range_ >= phys_step_);
```
Comment (sethrj) [inline src/celeritas/em/distribution/UrbanMscStepLimit.hh:148]: In what case isn't this true? I think we discussed this before: isn't it better to have an assertion than to add a `min` (effectively a branch in terms of testing requirements)?

Comment (whokion) [reply src/celeritas/em/distribution/UrbanMscStepLimit.hh:148]: If ionization is inactivated, the condition does not work as intended.  So it should not be assumed that the multiple scattering tightly depends on the existence of ionization process (even though msc needs the range information).

File: src/celeritas/em/distribution/UrbanMscScatter.hh [hunk 2]
```diff
@@ -192,6 +192,11 @@ CELER_FUNCTION auto UrbanMscScatter::operator()(Engine& rng) -> MscInteraction
         return MscInteraction{true_path_, inc_direction_, {0, 0, 0}};
     }
 
+    // Sample polar angle
+    real_type costheta
+        = this->sample_cos_theta(rng, input_.true_path, input_.limit_min);
+    CELER_ASSERT(std::fabs(costheta) <= 1);
+
```
[the remark this answers is not in the document]
Comment (whokion) [reply src/celeritas/em/distribution/UrbanMscScatter.hh:199]: inside the `sample_cos_theta`, t`au_` will be calculated of which updated information needs for evaluating the lateral displacement.

File: src/celeritas/em/distribution/UrbanMscStepLimit.hh (hunk 1 above)
Comment (sethrj) [reply src/celeritas/em/distribution/UrbanMscStepLimit.hh:148]: This might be how Geant works, but we will be able to define an energy loss/range term without requiring an ionization process. In fact our current implementation we always have range limiter if there's an eloss process, and msc requires it. So this "min" outside of your unit test is never needed. I'd much rather the "min" be part of your unit test.

File: src/celeritas/em/distribution/UrbanMscScatter.hh (hunk 2 above)
[the remark this answers is not in the document]
Comment (sethrj) [reply src/celeritas/em/distribution/UrbanMscScatter.hh:199]: Ughh side effects! Good catch. Please add a comment about the ordering.

File: src/celeritas/em/distribution/UrbanMscStepLimit.hh (hunk 1 above)
Comment (whokion) [reply src/celeritas/em/distribution/UrbanMscStepLimit.hh:148]: It still does not cover the case that the trial physics step is smaller than the range which is legitimated request to the step limit (otherwise, the check is equivalent passing the range instead of physics step in the arugment) - in general, we should not decide what physics procesees should be in or tested.

File: src/celeritas/em/distribution/UrbanMscScatter.hh (hunk 2 above)
[the remark this answers is not in the document]
Comment (whokion) [reply src/celeritas/em/distribution/UrbanMscScatter.hh:199]: Added a comment.

File: src/celeritas/em/distribution/UrbanMscStepLimit.hh (hunk 1 above)
Comment (sethrj) [reply src/celeritas/em/distribution/UrbanMscStepLimit.hh:148]: Wait, so unless I'm mistaken "phys step" here is the most limiting distance: discrete interaction or range at this point. Are you expecting it to be *only* the discrete step limit?

Comment (whokion) [reply src/celeritas/em/distribution/UrbanMscStepLimit.hh:148]: No. What if one wants to simulate a charged particle scattering without the energy loss process regardless the test is physically wrong or right? - this is exactly what TestEM15 does. Should a MSC model support such a case?  My answer is yes as I, as an end user have a way to configure or control any envisaged physical simulation for what ever reason. If we impose the constraint (range> physics step) in the constructor, we are assuming that MSC always comes with the energy loss process (i.e, we can combine them together as a unit) - it is physically a correct assumption in standard cases, but rules out other potential studies - for an example, what if a charged particle is accelerated in the electric field so that it can fly longer than the dEdx range or one want to only study an impact of a MSC model.  Entangling a presumed entire workflow in an efficient way is important, but we also should provide all model-level simulation independently.  I do not think that the extra branch is a potential performance killer as the test is always true for the standard work flow.

**Comment by whokion** [comment]
@sethrj the test (celeritas_em_UrbanMsc) was successful on my side (on two differently platforms). Can you pass any useful error message from /var/jenkins/workspace/Celeritas_PR-429/build/Testing/Temporary/LastTest.log?  Thanks.

File: src/celeritas/em/distribution/UrbanMscStepLimit.hh (hunk 1 above)
Comment (sethrj) [reply src/celeritas/em/distribution/UrbanMscStepLimit.hh:148]: > What if one wants to simulate a charged particle scattering without the energy loss process regardless the test is physically wrong or right?

Since the MSC classes themselves *require* dE/dx data, this isn't possible, is it? And if you have a dE/dx, then isn't there a maximum step/range implied by that data and the particle's start energy? I get what you mean as a user, but is it useful to test the result of using a "step length" that exceeds that range? How will we know that other constraints in the code aren't violated by that result (e.g. losing more energy than the particle starts witih).

> I do not think that the extra branch is a potential performance killer as the test is always true for the standard work flow.

I'm not at all worried about performance here but rather code complexity and readability. Adding a `min` means you should make sure the code is tested for both "branches" of that minimum operation, and worse it makes it harder to reason about the program flow. With an assertion, you *know* what the values are supposed to be. With a min/max/whatever, you're saying "this can be anything but if one value is larger than the other we'll choose the smaller one" -- it doesn't say *which* value you expect to be chosen in the typical case or *why* it might be violated.

The code complexity is what gets me most: if this "min" ends up silently allowing some other downstream failure to happen, it makes it harder to know that "oh that actually shouldn't have been possible there". I'd feel much better if you can document what the two cases of `min` are, but I still don't like that we're turning a strong guarantee into a weak one.

Comment (whokion) [reply src/celeritas/em/distribution/UrbanMscStepLimit.hh:148]: The dEdx data table can be built (so, can be used in MSC), but the ionization model can be inactivated (i.e., not participating the GPIL competition). This is what Geant4 TestEM15 does - sure that we do not need to follow what Geant4 supports, but needs a good reason why we rule out all other use cases other than the demo-loop.  I will close this MR, and will open another one with just bug fixes.

Comment (sethrj) [reply src/celeritas/em/distribution/UrbanMscStepLimit.hh:148]: The range calculation and GPIL for ionization are separate in celeritas, that's the thing..:

Comment (whokion) [reply src/celeritas/em/distribution/UrbanMscStepLimit.hh:148]: Same in Geant4  - i.e, PostStepGPIL (ioni) and AlongStepGPIL (range).  Anyway, let's move on as I closed this MR and open another one with necessary bug fixes which should give same physics results with respect to TestEM15.
